diff --git a/clib/hashset.ml b/clib/hashset.ml
index 224f870..e07b578 100644
--- a/clib/hashset.ml
+++ b/clib/hashset.ml
@@ -11,6 +11,8 @@
 (** Adapted from Damien Doligez, projet Para, INRIA Rocquencourt,
     OCaml stdlib. *)
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+
 (** The following functor is a specialized version of [Weak.Make].
     Here, the responsibility of computing the hash function is now
     given to the caller, which makes possible the interleaving of the
@@ -230,9 +232,9 @@ module Combine = struct
        this topic. Therefore, there must be room for improvement here. *)
     let alpha = 65599
     let beta  = 7
-    let combine x y     = x * alpha + y
+    let combine x y     = (x * alpha + y) land 0x3fffffff  (* coerce-32bit.patch *)
     let combine3 x y z   = combine x (combine y z)
     let combine4 x y z t = combine x (combine3 y z t)
     let combine5 x y z t u = combine x (combine4 y z t u)
-    let combinesmall x y = beta * x + y
+    let combinesmall x y = (beta * x + y) land 0x3fffffff  (* coerce-32bit.patch *)
 end
diff --git a/kernel/dune b/kernel/dune
index d5560ce..3f8a5c9 100644
--- a/kernel/dune
+++ b/kernel/dune
@@ -25,12 +25,12 @@
 
 (rule
  (targets uint63.ml)
- (deps (:gen-file uint63_%{ocaml-config:int_size}.ml))
+ (deps (:gen-file uint63_31.ml))  ; (* coerce-32bit.patch *)
  (action (copy# %{gen-file} %{targets})))
 
 (rule
  (targets float64.ml)
- (deps (:gen-file float64_%{ocaml-config:int_size}.ml))
+ (deps (:gen-file float64_31.ml))  ; (* coerce-32bit.patch *)
  (action (copy# %{gen-file} %{targets})))
 
 (documentation
diff --git a/kernel/nativecode.ml b/kernel/nativecode.ml
index eb78020..eb18943 100644
--- a/kernel/nativecode.ml
+++ b/kernel/nativecode.ml
@@ -19,6 +19,8 @@ open Genlambda
 open Nativelambda
 open Environ
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+
 (** This file defines the mllambda code generation phase of the native
 compiler. mllambda represents a fragment of ML, and can easily be printed
 to OCaml code. *)
diff --git a/kernel/uint63_31.ml b/kernel/uint63_31.ml
index 1c6a448..10168d7 100644
--- a/kernel/uint63_31.ml
+++ b/kernel/uint63_31.ml
@@ -11,7 +11,7 @@
 (* Invariant: the msb should be 0 *)
 type t = Int64.t
 
-let _ = assert (Sys.word_size = 32)
+(* let _ = assert (Sys.word_size = 32) *)    (* coerce-32bit.patch *)
 
 let uint_size = 63
 
diff --git a/lib/objFile.ml b/lib/objFile.ml
index 1c742d2..b65dc2b 100644
--- a/lib/objFile.ml
+++ b/lib/objFile.ml
@@ -116,7 +116,7 @@ let marshal_out_segment h ~segment v =
   let { out_channel = ch } = h in
   let () = assert (not (CString.Map.mem segment.id h.out_segments)) in
   let pos = LargeFile.pos_out ch in
-  let () = Marshal.to_channel ch v [] in
+  let () = Marshal.to_channel ch v [Marshal.Compat_32] in    (* coerce-32bit.patch *)
   let () = flush ch in
   let pos' = LargeFile.pos_out ch in
   let len = Int64.sub pos' pos in
diff --git a/lib/system.ml b/lib/system.ml
index c6a370e..63ae23d 100644
--- a/lib/system.ml
+++ b/lib/system.ml
@@ -252,7 +252,7 @@ let check_caml_version ~caml:s ~file:f =
     be compatible.")
   else ()
 
-let marshal_out ch v = Marshal.to_channel ch v []; flush ch
+let marshal_out ch v = Marshal.to_channel ch v [Marshal.Compat_32]; flush ch   (* coerce-32bit.patch *)
 let marshal_in filename ch =
   try Marshal.from_channel ch
   with
diff --git a/plugins/extraction/mlutil.ml b/plugins/extraction/mlutil.ml
index 94edbd9..48c4e31 100644
--- a/plugins/extraction/mlutil.ml
+++ b/plugins/extraction/mlutil.ml
@@ -16,6 +16,8 @@ open Table
 open Miniml
 (*i*)
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+
 (*s Exceptions. *)
 
 exception Found
diff --git a/plugins/ltac/tacentries.ml b/plugins/ltac/tacentries.ml
index 5f663a2..71564fd 100644
--- a/plugins/ltac/tacentries.ml
+++ b/plugins/ltac/tacentries.ml
@@ -239,7 +239,7 @@ let make_fresh_key =
     (* We embed the hash of the kernel name in the label so that the identifier
        should be mostly unique. This ensures that including two modules
        together won't confuse the corresponding labels. *)
-    let hash = (cur lxor (ModPath.hash (Lib.current_mp ()))) land 0x7FFFFFFF in
+    let hash = (cur lxor (ModPath.hash (Lib.current_mp ()))) land 0x3FFFFFFF in   (* coerce-32bit.patch *)
     let lbl = Id.of_string_soft (Printf.sprintf "%s_%08X" prods hash) in
     Lib.make_kn lbl
 
diff --git a/plugins/ltac2/tac2entries.ml b/plugins/ltac2/tac2entries.ml
index 96e6ddc..5390c2f 100644
--- a/plugins/ltac2/tac2entries.ml
+++ b/plugins/ltac2/tac2entries.ml
@@ -813,7 +813,7 @@ let make_fresh_key tokens =
   (* We embed the hash of the kernel name in the label so that the identifier
       should be mostly unique. This ensures that including two modules
       together won't confuse the corresponding labels. *)
-  let hash = (ModPath.hash (Lib.current_mp ())) land 0x7FFFFFFF in
+  let hash = (ModPath.hash (Lib.current_mp ())) land 0x3FFFFFFF in   (* coerce-32bit.patch *)
   let lbl = Id.of_string_soft (Printf.sprintf "%s_%08X" prods hash) in
   Lib.make_kn lbl
 
diff --git a/plugins/micromega/coq_micromega.ml b/plugins/micromega/coq_micromega.ml
index 5558a7e..a933478 100644
--- a/plugins/micromega/coq_micromega.ml
+++ b/plugins/micromega/coq_micromega.ml
@@ -29,6 +29,8 @@ open McPrinter
 
 module ERelevance = EConstr.ERelevance
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+
 (**
   * Debug flag
   *)
diff --git a/plugins/micromega/linsolve.ml b/plugins/micromega/linsolve.ml
index 6d9e14a..1d3464f 100644
--- a/plugins/micromega/linsolve.ml
+++ b/plugins/micromega/linsolve.ml
@@ -9,6 +9,9 @@
 (************************************************************************)
 
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+let min_int = -(1 lsl 30)     (* coerce-32bit.patch *)
+
 (** An equation is of the following form a1.x1 + a2.x2 = c
 *)
 
diff --git a/plugins/micromega/persistent_cache.ml b/plugins/micromega/persistent_cache.ml
index 6577f93..f831bb5 100644
--- a/plugins/micromega/persistent_cache.ml
+++ b/plugins/micromega/persistent_cache.ml
@@ -160,7 +160,7 @@ module PHashtable (Key : HashedType) : PHashtable with type key = Key.t = struct
   let add t k e =
     let {outch} = t in
     let fd = descr_of_out_channel outch in
-    let h = Key.hash k land 0x7FFFFFFF in
+    let h = Key.hash k land 0x3FFFFFFF in    (* coerce-32bit.patch *)
     let dump () =
       let () = output_binary_int outch h in
       let () = Marshal.to_channel outch (k, e) [] in
@@ -172,7 +172,7 @@ module PHashtable (Key : HashedType) : PHashtable with type key = Key.t = struct
 
   let find t k =
     let {outch; htbl = tbl} = t in
-    let h = Key.hash k land 0x7FFFFFFF in
+    let h = Key.hash k land 0x3FFFFFFF in    (* coerce-32bit.patch *)
     let lpos = Table.find h tbl in
     (* First look for already live data *)
     let find data = match data.obj with
diff --git a/plugins/micromega/polynomial.ml b/plugins/micromega/polynomial.ml
index 1a61e34..9e0f739 100644
--- a/plugins/micromega/polynomial.ml
+++ b/plugins/micromega/polynomial.ml
@@ -19,6 +19,8 @@ open Q.Notations
 open Mutils
 module Mc = Micromega
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+
 let max_nb_cstr = ref max_int
 
 type var = int
diff --git a/vernac/metasyntax.ml b/vernac/metasyntax.ml
index da65176..21b6073 100644
--- a/vernac/metasyntax.ml
+++ b/vernac/metasyntax.ml
@@ -27,6 +27,8 @@ open Libnames
 open Notation
 open Nameops
 
+let max_int = (1 lsl 30) - 1  (* coerce-32bit.patch *)
+
 (** **************************************************************** **)
 (** Printing grammar entries                                         **)
 
